graphrag/query/structured_search/drift_search/state.py

```diff
@@ -0,0 +1,141 @@
+# Copyright (c) 2024 Microsoft Corporation.
+# Licensed under the MIT License
+
+"""Manage the state of the DRIFT query, including a graph of actions."""
+
+import logging
+import random
+from collections.abc import Callable
+from typing import Any
+
+import networkx as nx
+
+from graphrag.query.structured_search.drift_search.action import DriftAction
+
+log = logging.getLogger(__name__)
+
+
+class QueryState:
+    """Manage the state of the query, including a graph of actions."""
+
+    def __init__(self):
+        self.graph = nx.MultiDiGraph()
+
+    def add_action(self, action: DriftAction, metadata: dict[str, Any] | None = None):
+        """Add an action to the graph with optional metadata."""
+        self.graph.add_node(action, **(metadata or {}))
+
+    def relate_actions(
+        self, parent: DriftAction, child: DriftAction, weight: float = 1.0
+    ):
+        """Relate two actions in the graph."""
+        self.graph.add_edge(parent, child, weight=weight)
+
+    def add_all_follow_ups(
+        self,
+        action: DriftAction,
+        follow_ups: list[DriftAction] | list[str],
+        weight: float = 1.0,
+    ):
+        """Add all follow-up actions and links them to the given action."""
+        if len(follow_ups) == 0:
+            log.warning("No follow-up actions for action: %s", action.query)
+
+        for follow_up in follow_ups:
+            if isinstance(follow_up, str):
```

Comment:
According to the method signature it should be a list of DriftActions, not str

Reply:
nvm, didn't see the | list[str]. but still, if the type is the first one, list[DriftAction] and it throws a warning, why allow it?

Reply:
Yup, fixing that actually